Behavior-Driven Development Reloaded©

domi41 opened the pull request:
See #103 
See #198 

---

~~Only one~~ Two features file have been ported from #198 (the others fail because not only the codebase changed significantly, the specs themselves have changed and need to be rewritten).

The provided features are about solving issue #201 and authoring a new idea.

Still, the feature suite runs and is ready for more.

domi41 commented on features/authoring-ideas.feature:
```diff
@@ -0,0 +1,30 @@
+      And I type "Require libre software in all governmental endeavours."
+        # Here, if we do not wait, an idea without title is submitted.
+        # Not sure what this issue is about nor how to solve it. Help?
+      And I wait for 1 second
```
This is really hacky and will probably cause trouble on CI, as `wait` instructions may not be long enough there. It can be fixed (probably), but needs help.

vjrj replied:
Yes, this is hacky. Are you testing that input fields are focusing correctly? I'll test this PR after the rebase.

If I understand well this feature, maybe you can use [hasFocus](http://webdriver.io/api/state/hasFocus.html) instead, and a feature like `And the title get the focus`.

I didn't test if something is focused previously. I usually just type in some field and use 'waitForVisible' and/or 'waitForText'. You can see a sample in [this other meteor app](https://github.com/comunes/todos-contra-el-fuego-web/blob/a6bd1095bf79e1f376f9b27abfbb2fd6ecdf03ea/cucumber/features/steps_definitions/login.js#L35) or login/register in the app.

domi41 replied:
Thanks for the feedback and the useful link ! :heart_eyes: This one is not about focus, nor about waiting for an element (though we can perhaps solve this by waiting for the right DOM element). It seems related to some meteor voodoo, possibly also because we're using `contenteditable` instead of `input`. We're still trudging through the changes since the last rebase, but it's coming along.

domi41 commented:
Everything is broken because of the usage of `Session` during user creation. (in [transaction.js](https://github.com/DemocracyEarth/sovereign/blob/a63db7dd25276536ebb48bd650b78743f7121454/imports/api/transactions/transaction.js#L415))

We're hacking blindly around to try and make it work without `Session`. Is it really necessary ? It precludes server-side creation of users, and we need that.
We found this : https://github.com/matteodem/meteor-server-session

domi41 commented:
Creating a new user now creates an idea (a Contract) for this user. We did **not** expect this.
Is that nominal behavior, or did we break something ? @santisiri 

``` gherkin
  Scenario: Propose an idea successfully
  ✔ Given there should be zero ideas in the database
  ✔ And I dump all the ideas
#   | []
  ✔ Given I am the citizen named Richard S.
  ✔ And I dump all the ideas
    | [
    | 	{
    | 		"_id": "4nPqGq2PqLJPBprnn",
    | 		"keyword": "draft-a42CeRnYTDJz3LW2s",
    | 		"collectiveId": "JNJ2LXHTAc4sLLKLY",
    | 		"title": "",
    | 		"kind": "VOTE",
    | 		"context": "GLOBAL",
    | 		"url": "/vote/",
    | 		"description": "",
    | 		"createdAt": "2018-04-04T18:09:15.186Z",
    | 		"lastUpdate": "2018-04-04T18:09:15.201Z",
    | 		"timestamp": "2018-04-04T18:09:15.201Z",
    | 		"tags": [],
    | 		"membersOnly": false,
    | 		"permanentElection": true,
    | 		"executionStatus": "OPEN",
    | 		"anonymous": false,
    | 		"closingDate": "2018-04-05T18:09:15.187Z",
    | 		"alwaysOpen": false,
    | 		"allowForks": false,
    | 		"secretVotes": false,
    | 		"realtimeResults": false,
    | 		"multipleChoice": false,
    | 		"rankPreferences": false,
    | 		"executiveDecision": true,
    | 		"stage": "DRAFT",
    | 		"ballot": [],
    | 		"ballotEnabled": true,
    | 		"authorized": false,
    | 		"isDefined": false,
    | 		"isRoot": true,
    | 		"events": [],
    | 		"wallet": {
    | 			"balance": 0,
    | 			"placed": 0,
    | 			"available": 0,
    | 			"currency": "VOTES",
    | 			"address": [],
    | 			"ledger": []
    | 		},
    | 		"signatures": [
    | 			{
    | 				"_id": "a42CeRnYTDJz3LW2s",
    | 				"role": "AUTHOR",
    | 				"hash": "",
    | 				"username": "richards",
    | 				"status": "CONFIRMED"
    | 			}
    | 		]
    | 	}
    | ]
```

domi41 commented:
Ok, thanks ! We adjusted the suite accordingly, and added a scenario to make this optimization explicit.

``` gherkin
  Scenario: Create an idea draft upon user creation
  ✔ Given I am the citizen named Richard S.
  ✔ Then there should be one idea draft in the database
```

---

The suite now runs and is ready for a more thorough review.

domi41 commented on features/README.md:
```diff
@@ -0,0 +1,109 @@
+## WARNING
+
+Running the feature suite will TRUNCATE THE DATABASE before each Scenario.
+You WILL LOSE EVERYTHING on your development environment database.
```
I'm 100% for using a `test_` database, but I have not managed to do so yet. Keep in mind that I'm very new to Meteor and server-side javascript. I'm not too fond of javascript because it violates elementary rules like the transitivity of equality and the commutativity of addition, but that's a sore subject for another debate… If we can manage to configure the test suite to use its own database, let's do it !

vjrj replied:
Great!, some new script in `package.json` like:
```json
    "start:test-chimp": "MONGO_URL=mongodb://localhost:27017/sovereigntestchimp meteor --settings=config/development/settings.json",
```
works great:
```
3 scenarios (3 passed)
29 steps (29 passed)
```
using `sovereigntestchimp` specific database, so we can trash it without use the default dev database instance.

vjrj replied:
Yes, I prefer to add it to the PR, but up to you.

Anyway, this works for me:
```json
    "start:for-features": "MONGO_URL=mongodb://localhost:27017/sovereingtestchimp meteor --settings=config/development/settings.json --port 3003",
    "test:features": "./node_modules/.bin/chimp --chai --ddp=http://localhost:3003",
```
maybe your error it's because of your npm version? Here:
```bash
$ npm -v
5.6.0
$ meteor npm -v
5.6.0
```

I don't know how to do it in one line :-/ I usually run meteor in background in our CI, sleep 120, and later I run the tests that depends on it. Finally I stop meteor. Maybe I can improved it looking for the meter port in a loop:
```bash
# this inside a while til some timeout or 3003 start to listen
   lsof -i:3003 > /dev/null 2>&1
   # check if $? equal 0, so 3003 is listening, so we can start our tests
```

domi41 replied:
Ok, so it's a problem on my end, thanks for the tip !

```
$ npm -v
3.5.2
$ meteor npm -v
5.6.0
```

---

I'm investigating, but in case it's an obvious problem and you have the solution handy, I'll just dump here what I've got :

``` bash
$ meteor npm run test:features

> Sovereign.app@ test:features /home/domi41/code/snd/sovereign
> chimp --chai --ddp=http://localhost:3003

(node:4986) [DEP0022] DeprecationWarning: os.tmpDir() is deprecated. Use os.tmpdir() instead.

Master Chimp and become a testing Ninja! Check out our course: http://bit.ly/2btQaFu


[chimp] Running...
## There is an issue with `node-fibers` ##
`/home/domi41/code/snd/sovereign/node_modules/cucumber/node_modules/fibers/bin/linux-x64-57/fibers.node` is missing.

Try running this to fix the issue: /home/domi41/.meteor/packages/meteor-tool/.1.6.1.89za4a.ppaex++os.linux.x86_64+web.browser+web.cordova/mt-os.linux.x86_64/dev_bundle/bin/node /home/domi41/code/snd/sovereign/node_modules/cucumber/node_modules/fibers/build
/home/domi41/code/snd/sovereign/node_modules/cucumber/node_modules/fibers/fibers.js:20
	throw new Error('Missing binary. See message above.');
	^

Error: Missing binary. See message above.
    at Object.<anonymous> (/home/domi41/code/snd/sovereign/node_modules/cucumber/node_modules/fibers/fibers.js:20:8)
    at Module._compile (module.js:643:30)
    at Object.Module._extensions..js (module.js:654:10)
    at Module.load (module.js:556:32)
    at tryModuleLoad (module.js:499:12)
    at Function.Module._load (module.js:491:3)
    at Module.require (module.js:587:17)
    at require (internal/module.js:11:18)
    at Object.<anonymous> (/home/domi41/code/snd/sovereign/node_modules/cucumber/lib/cucumber/util/run.js:1:75)
    at Module._compile (module.js:643:30)
Cucumber steps failed
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! Sovereign.app@ test:features: `chimp --chai --ddp=http://localhost:3003`
npm ERR! Exit status 2
npm ERR! 
npm ERR! Failed at the Sovereign.app@ test:features script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/domi41/.npm/_logs/2018-04-09T16_45_47_914Z-debug.log


domi41@idao:~/code/snd/sovereign$ /home/domi41/.meteor/packages/meteor-tool/.1.6.1.89za4a.ppaex++os.linux.x86_64+web.browser+web.cordova/mt-os.linux.x86_64/dev_bundle/bin/node /home/domi41/code/snd/sovereign/node_modules/cucumber/node_modules/fibers/build
gyp info it worked if it ends with ok
gyp info using node-gyp@3.4.0
gyp info using node@6.11.4 | linux | x64
gyp info spawn /usr/bin/python2
gyp info spawn args [ '/usr/share/node-gyp/gyp/gyp_main.py',
gyp info spawn args   'binding.gyp',
gyp info spawn args   '-f',
gyp info spawn args   'make',
gyp info spawn args   '-I',
gyp info spawn args   '/home/domi41/code/snd/sovereign/build/config.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '/usr/share/node-gyp/addon.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '/usr/include/nodejs/common.gypi',
gyp info spawn args   '-Dlibrary=shared_library',
gyp info spawn args   '-Dvisibility=default',
gyp info spawn args   '-Dnode_root_dir=/usr/include/nodejs',
gyp info spawn args   '-Dnode_gyp_dir=/usr/share/node-gyp',
gyp info spawn args   '-Dnode_lib_file=node.lib',
gyp info spawn args   '-Dmodule_root_dir=/home/domi41/code/snd/sovereign',
gyp info spawn args   '--depth=.',
gyp info spawn args   '--no-parallel',
gyp info spawn args   '--generator-output',
gyp info spawn args   'build',
gyp info spawn args   '-Goutput_dir=.' ]
gyp: binding.gyp not found (cwd: /home/domi41/code/snd/sovereign) while trying to load binding.gyp
gyp ERR! configure error 
gyp ERR! stack Error: `gyp` failed with exit code: 1
gyp ERR! stack     at ChildProcess.onCpExit (/usr/share/node-gyp/lib/configure.js:304:16)
gyp ERR! stack     at emitTwo (events.js:106:13)
gyp ERR! stack     at ChildProcess.emit (events.js:191:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:219:12)
gyp ERR! System Linux 4.13.0-38-generic
gyp ERR! command "/usr/bin/node" "/usr/bin/node-gyp" "rebuild" "--release"
gyp ERR! cwd /home/domi41/code/snd/sovereign
gyp ERR! node -v v6.11.4
gyp ERR! node-gyp -v v3.4.0
gyp ERR! not ok 
node-gyp exited with code: 1
Please make sure you are using a supported platform and node version. If you
would like to compile fibers on this machine please make sure you have setup your
build environment--
Windows + OS X instructions here: https://github.com/nodejs/node-gyp
Ubuntu users please run: `sudo apt-get install g++ build-essential`
Alpine users please run: `sudo apk add python make g++`

```

Note that running the command by itself works, so it's probably a version mismatch problem, as you suspected.

---

And finally, since speed is not an issue during CI, if a sleep works I'm not against using that until we cook up a more complex bash script. One thing at a time ; the first thing is to start using BDD. Then, we can do it right :angel:

vjrj commented on .meteor/packages:
```diff
@@ -28,7 +31,7 @@ react-template-helper
 
 # data
-mongo@1.4.2                  # The database Meteor supports right now
+mongo@1.4.5                   # The database Meteor supports right now
```
Are all this package version changes necessary for this PR?

domi41 replied:
Nope, I just got lazy :see_no_evil:. Fixing this now… Thanks ;)

vjrj commented on .meteor/release:
```diff
@@ -1 +1 @@
-METEOR@1.6.1
+METEOR@1.6.1.1
```
Same here...

vjrj commented on package.json:
```diff
@@ -24,37 +24,43 @@
   },
   "dependencies": {
-    "@babel/runtime": "^7.0.0-beta.38",
+    "@babel/runtime": "^7.0.0-beta.44",
```
Same here...  I prefer to have a clean PR with the minimal packages changes related to this PR (without unnecessary meteor or npm packages upgrades). We can do a newer PR with meteor/npm upgrades, so if we have any problem with some commit in, for instance, production, we can identify correctly when/why something stopped to work, or a new bug arise, etc...

medied commented:
Happy to see this PR being brought back to life, good job @domi41 and @vjrj 👏

vjrj commented on .meteor/versions:
```diff
@@ -1,20 +1,20 @@
 aldeed:schema-deny@1.1.0
 aldeed:schema-index@1.1.1
-aldeed:simple-schema@1.5.3
+aldeed:simple-schema@1.5.4
```
Thanks @domi41 for the PR update. I see here the same issue: Many package changes that, if I'm not wrong, are not related with this PR, so we can introduce unexpected bugs. If you have some more spare time can you clean up more `.meteor/versions`?

virgile-dev commented:
Great news to see this was merged !
